include/ginkgo/core/base/collection.hpp

```diff
@@ -0,0 +1,200 @@
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+#include <numeric>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/array.hpp>
+
+
+namespace gko {
+namespace collection {
+
+/**
+ * \brief A minimal std::vector-like interface for a collection of arrays.
+ *
+ * The arrays are stored as sub-views of a single contiguous array.
+ *
+ * \tparam T value type stored in the arrays
+ */
+template <typename T>
+struct array {
+    using value_type = T;
+    using iterator = gko::array<T>*;
+    using const_iterator = const gko::array<T>*;
+    using reference = gko::array<T>&;
+    using const_reference = const gko::array<T>&;
+
+    array() = default;
+
+    explicit array(std::shared_ptr<const Executor> exec)
+        : buffer_(exec), offsets_(exec)
+    {}
+
+    array(std::shared_ptr<const Executor> exec, const array& other)
+        : array(exec)
+    {
+        *this = other;
+    }
+
+    array(std::shared_ptr<const Executor> exec, array&& other) : array(exec)
+    {
+        *this = std::move(other);
+    }
+
+    array(const array& other) { *this = other; }
+
+    array& operator=(const array& other)
+    {
+        if (this != &other) {
+            buffer_ = other.buffer_;
+            offsets_ = other.offsets_;
+            elems_.resize(other.elems_.size());
+            for (size_type i = 0; i < elems_.size(); ++i) {
+                elems_[i] = make_array_view(
+                    buffer_.get_executor(), other[i].get_size(),
+                    buffer_.get_data() +
+                        std::distance(other.buffer_.get_const_data(),
+                                      other[i].get_const_data()));
+            }
+        }
+        return *this;
+    }
+
+    // can't be noexcept, because array is not noexcept
+    array(array&&) = default;
+
+    array& operator=(array&&) = default;
+
+    template <typename SizeType = std::size_t>
+    array(std::shared_ptr<const Executor> exec,
+          const std::vector<SizeType>& sizes)
+        : array(gko::array<T>(exec,
+                              std::accumulate(sizes.begin(), sizes.end(), 0)),
+                sizes)
+    {}
+
+    template <typename SizeType = std::size_t>
+    array(gko::array<T> buffer, const std::vector<SizeType>& sizes)
+        : buffer_(std::move(buffer))
+    {
+        auto exec = buffer_.get_executor();
+        offsets_ = gko::array<size_type>(exec->get_master(), sizes.size() + 1);
+        offsets_.fill(0);
+        std::partial_sum(sizes.begin(), sizes.end(), offsets_.get_data() + 1);
+        buffer_.resize_and_reset(offsets_.get_data()[sizes.size()]);
+        // can't use emplace/push_back because then the view will get copied
+        // into owning arrays
+        elems_.resize(sizes.size());
+        for (size_type i = 0; i < sizes.size(); ++i) {
+            elems_[i] = make_array_view(
+                exec, sizes[i], buffer_.get_data() + offsets_.get_data()[i]);
+        }
```

Comment:
I think this design might encourage poor performance - it invites using a kernel invocation for each MPI rank, which might scale poorly. Instead I would try to do this with a single kernel for all ranks, and a corresponding device view.